mods/custom-desktop-watermark.wh.cpp

```diff
@@ -0,0 +1,201 @@
+{
+    return true;
+}
+
+int PaintLine(
+    HDC            hDC,
+    LPCRECT        lprc,
+    LPCWSTR        lpszText,
+    HFONT          hFont,
+    int            offset
+)
+{
+    HFONT hfOld = (HFONT)SelectObject(hDC, hFont);
+        
+    RECT rcText = { 0 };
+    int newOffset = DrawTextW(
+        hDC, lpszText, -1,
+        &rcText, DT_CALCRECT | DT_SINGLELINE
+    );
+
+    RECT rcPaint = { 0 };
+    rcPaint.left = lprc->right - RECTWIDTH(rcText);
+    rcPaint.top = lprc->bottom - RECTHEIGHT(rcText);
+    rcPaint.right = lprc->right;
+    rcPaint.bottom = lprc->bottom;
+
+    int padding = MulDiv(5, GetDeviceCaps(hDC, LOGPIXELSX), 96);
+    OffsetRect(&rcPaint, -padding, -offset);
+
+    DrawTextW(hDC, lpszText, -1, &rcPaint, DT_SINGLELINE);
+
+    SelectObject(hDC, hfOld);
+    return newOffset;    
+}
+
+void (*CDesktopWatermark_s_DesktopBuildPaint_orig)(HDC, LPCRECT, HFONT);
+void CDesktopWatermark_s_DesktopBuildPaint_hook(
+    HDC     hDC,
+    LPCRECT lprc,
+    HFONT   hFont
+)
+{
+    COLORREF cr = SetTextColor(hDC, RGB(255, 255, 255));
+    int offset = 0;
+
+    NONCLIENTMETRICSW ncm = { sizeof(ncm) };
+    int dpi = GetDeviceCaps(hDC, LOGPIXELSX);
+    SystemParametersInfoForDpi(
+        SPI_GETNONCLIENTMETRICS,
+        sizeof(ncm),
+        &ncm,
+        0,
+        dpi
+    );
+    HFONT hTitleFont = CreateFontIndirectW(&ncm.lfCaptionFont);
+    HFONT hMessageFont = CreateFontIndirectW(&ncm.lfMessageFont);
+    int padding = MulDiv(3, dpi, 96);
+    offset += MulDiv(g_bClassic ? 4 : 1, dpi, 96);
+
+    for (size_t i = g_lines.size(); i--;)
+    {
+        WatermarkLine line = g_lines.at(i);
+        bool bMessageFont = g_bClassic && !line.title;
+
+        offset += PaintLine(
+            hDC, lprc, line.text.c_str(),
+            bMessageFont ? hMessageFont : hTitleFont,
+            offset
+        ) + padding;
+    }
+
+    SetTextColor(hDC, cr);
+}
+
+const WindhawkUtils::SYMBOL_HOOK shell32DllHooks[] = {
+    {
+        {
+            L"public: static bool __cdecl CDesktopWatermark::s_WantWatermark(void)"
+        },
+        &CDesktopWatermark_s_WantWatermark_orig,
+        CDesktopWatermark_s_WantWatermark_hook,
+        false
+    },
+    {
+        {
+            L"private: static void __cdecl CDesktopWatermark::s_DesktopBuildPaint(struct HDC__ *,struct tagRECT const *,struct HFONT__ *)"
+        },
+        &CDesktopWatermark_s_DesktopBuildPaint_orig,
+        CDesktopWatermark_s_DesktopBuildPaint_hook,
+        false
+    }
+};
+
+void LoadSettings(void)
+{
+    g_lines.clear();
+    g_bClassic = Wh_GetIntSetting(L"classic");
+
+    for (int i = 0;; i++)
+    {
+        LPCWSTR lpszText = Wh_GetStringSetting(L"lines[%i].text", i);
+        if (!lpszText || !*lpszText)
+            break;
+        bool title = Wh_GetIntSetting(L"lines[%i].title", i);
+        WatermarkLine line = { lpszText, title };
+        g_lines.push_back(line);
+        Wh_FreeStrintSetting(lpszText);
```

Comment:
Typo. Also:
* Not freeing when breaking at `!*lpszText` is still a leak.
* `Wh_GetStringSetting` never returns NULL. The check doesn't hurt, just FYI.